go-controller/pkg/network-attach-def-controller/network_attach_def_controller.go

```diff
@@ -260,10 +308,26 @@ func (nadController *NetAttachDefinitionController) syncNAD(key string, nad *net
 		return nil
 	}
 
-	// ensure the network associated with the NAD
 	ensureNetwork.AddNADs(key)
+
+	if ensureNetwork.IsPrimaryNetwork() {
+		nadController.primaryNetworks.LockKey(namespace)
+		var primaryNetInfos map[string]util.NetInfo
+		var loaded bool
+		if primaryNetInfos, loaded = nadController.primaryNetworks.Load(namespace); loaded {
+			primaryNetInfos[ensureNetwork.GetNetworkName()] = util.CopyNetInfo(ensureNetwork)
```

Comment:
In regards to this CopyNetInfo, I just checked that the NetInfo NAD methods are synchronized:

https://github.com/ovn-org/ovn-kubernetes/blob/master/go-controller/pkg/util/multi_network.go#L364-L397

and the rest is immutable. So I am not sure what kind of race you got there.

Reply:
I wish I had saved it. I also noticed that and was really confused at first how the race was happening. Then I figured it out and now I've forgotten it. I'll try to remember/reproduce it.

Reply:
So I managed to reproduce something. I'm guessing this is it but I cant be sure:

https://gist.github.com/trozet/1ebc50a7c4919f778c5c08c227f60c39

This looks like it is racing because the test is doing a reflect deepEqual which doesnt care about the netinfo mutex while the network manager controller is trying to update NADs and has the mutex.

Either way I think it is a bad idea to be propagating pointers between controller caches. I'd like to keep the copying of the netinfo in place unless you have an objection.

Reply:
Hmm okay, I agree that it might not be the best idea to be propagating pointers between controller caches. 

But I am not sure this specific copy will solve the issue. It looks like it is the test which should copy before doing deep equal.